src/Rings/MPolyMap/flattenings.jl

```diff
@@ -0,0 +1,247 @@
+########################################################################
+# Computation of kernels of homomorphisms of polynomial rings 
+# over coefficient rings which are not fields.
+#
+# In the admissible cases, this flattens the structure of the rings 
+# and computes the kernel on the flat counterpart. 
+########################################################################
+function kernel(
+    f::MPolyAnyMap{<:AbstractAlgebra.Generic.MPolyRing{<:RingElemType}, 
+                   <:AbstractAlgebra.Generic.MPolyRing{<:RingElemType}
+                  }
+  ) where {RingElemType <: Union{<:MPolyElem, <:MPolyQuoElem, <:MPolyLocalizedRingElem, 
+                                 <:MPolyQuoLocalizedRingElem
+                                }
+          }
+  f_flat = _flatten(f)
+  K_flat = kernel(f_flat)
+  K = ideal(domain(f), _flat_to_orig(domain(f)).(gens(K_flat)))
+  #set_attribute!(K, :_flat_counterpart, K_flat) # TODO: Support attribute storage?
+  return K
+end
+
+@attr function _flatten(
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: MPolyElem}
+  R = base_ring(S)
+  kk = coefficient_ring(R)
+  S_flat, _ = PolynomialRing(kk, vcat(symbols(S), symbols(R)))
+  R_to_S_flat = hom(R, S_flat, gens(S_flat)[ngens(S)+1:end])
+  S_to_S_flat = hom(S, S_flat, R_to_S_flat, gens(S_flat)[1:ngens(S)])
+  S_flat_to_S = hom(S_flat, S, vcat(gens(S), S.(gens(R))))
+  
+  set_attribute!(S, :_flat_to_orig, S_flat_to_S)
+  set_attribute!(S, :_orig_to_flat, S_to_S_flat)
+  set_attribute!(S, :_base_ring_to_flat, R_to_S_flat)
+
+  return S_flat
+end
+
+@attr function _flatten(
+    S::AbstractAlgebra.Generic.MPolyRing{RingElemType}
+  ) where {RingElemType <: MPolyQuoElem}
+  A = base_ring(S)::MPolyQuo
+  R = base_ring(A)::MPolyRing
+  I = modulus(A)::MPolyIdeal
+  kk = coefficient_ring(R)::Field
```

Comment:
Why these type annotations?

Reply:
These came out of writing the code and recalling what these objects would be, respectively, in order not to get too confused. Then, I just left them there, because they don't hurt and probably they even improve readability. 

Would you remove them?

Reply:
We don't use them for documentation purposes elsewhere, that is why I was wondering.